File: apps/api/app/services/push_contract.py
"""
Push wire contract — server side. The only place a payload is built.

Mirrors libs/shared/contracts/src/push-contract.ts — change both together.

Rules: every data value is a string (FCM requires it); `route` is server-decided
so retargeting a notification is not an app release; no order details, ever —
push renders on a locked screen in public, so ids and generic copy only.
"""
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone
from uuid import UUID

# Keep in sync with push-contract.ts
PUSH_CONTRACT_VERSION = "1"

TYPE_OFFER_PUBLISHED = "offer_published"
TYPE_OFFER_REVOKED = "offer_revoked"
TYPE_ORDER_ASSIGNED = "order_assigned"
TYPE_ORDER_UPDATED = "order_updated"

# Versioned: an Android channel's importance is immutable once created.
PUSH_CATEGORY_OFFER = "order-offers-v1"

ACTION_DISMISS = "dismiss"
ACTION_DETAILS = "details"


@dataclass(frozen=True, slots=True)
class PushEnvelope:
    data: dict[str, str]
    title: str
    body: str
    ttl_seconds: int
    expires_at: datetime
    category: str
    silent: bool


def build_offer_published(
    *,
    order_id: UUID | str,
    expires_at: datetime,
    now: datetime | None = None,
) -> PushEnvelope:
    """New broadcast offer, for on-shift drivers."""
    now = now or datetime.now(timezone.utc)
    ttl = max(0, int((expires_at - now).total_seconds()))
    minutes_left = max(1, ttl // 60)
    title = "New delivery offer"
    body = f"Tap for details - expires in {minutes_left} min"

    return PushEnvelope(
        data={
            "v": PUSH_CONTRACT_VERSION,
            "type": TYPE_OFFER_PUBLISHED,
            "order_id": str(order_id),
            "route": f"/offer/{order_id}",
            "title": title,
            "body": body,
            "expires_at": expires_at.isoformat(),
            "category": PUSH_CATEGORY_OFFER,
        },
        title=title,
        body=body,
        ttl_seconds=ttl,
        expires_at=expires_at,
        category=PUSH_CATEGORY_OFFER,
        silent=False,
    )


def build_offer_revoked(*, order_id: UUID | str, now: datetime | None = None) -> PushEnvelope:
    """Clear an offer notification already on other drivers' phones."""
    now = now or datetime.now(timezone.utc)
    return PushEnvelope(
        data={
            "v": PUSH_CONTRACT_VERSION,
            "type": TYPE_OFFER_REVOKED,
            "order_id": str(order_id),
            "route": "",
            "title": "",
            "body": "",
            "expires_at": now.isoformat(),
            "category": PUSH_CATEGORY_OFFER,
            "silent": "1",
        },
        title="",
        body="",
        ttl_seconds=300,
        expires_at=now,
        category=PUSH_CATEGORY_OFFER,
        silent=True,
    )


def build_order_assigned(
    *,
    order_id: UUID | str,
    order_number: str | None,
    expires_at: datetime,
    now: datetime | None = None,
) -> PushEnvelope:
    """Direct assignment. Not presence-gated — the work is already theirs."""
    now = now or datetime.now(timezone.utc)
    ttl = max(0, int((expires_at - now).total_seconds()))
    label = f"Order #{order_number}" if order_number else "A new order"
    title = "New job assigned"
    body = f"{label} has been assigned to you"

    return PushEnvelope(
        data={
            "v": PUSH_CONTRACT_VERSION,
            "type": TYPE_ORDER_ASSIGNED,
            "order_id": str(order_id),
            "route": f"/job/{order_id}",
            "title": title,
            "body": body,
            "expires_at": expires_at.isoformat(),
            "category": PUSH_CATEGORY_OFFER,
        },
        title=title,
        body=body,
        ttl_seconds=ttl,
        expires_at=expires_at,
        category=PUSH_CATEGORY_OFFER,
        silent=False,
    )

File: libs/shared/contracts/src/push-contract.ts
/**
 * Push wire contract. Mirrored server-side in
 * `apps/api/app/services/push_contract.py` — change both together; drift fails
 * silently on a driver's phone, not in a build.
 *
 * Rules:
 * 1. Every value is a string — FCM's `data` map is `Map<string, string>`.
 * 2. `route` is server-decided, so retargeting isn't an app release.
 * 3. Unknown `type`/`v` is ignored silently, so the server can add types first.
 */

/** Bumped only for a breaking payload change; unknown versions are ignored. */
export const PUSH_CONTRACT_VERSION = '1';

export const PUSH_TYPES = {
  OFFER_PUBLISHED: 'offer_published',
  OFFER_REVOKED: 'offer_revoked',
  ORDER_ASSIGNED: 'order_assigned',
  ORDER_UPDATED: 'order_updated',
} as const;

export type PushType = (typeof PUSH_TYPES)[keyof typeof PUSH_TYPES];

/**
 * Android channel id / iOS category id. Versioned because an Android channel's
 * importance is immutable once created — changing it needs a new id.
 */
export const PUSH_CATEGORY_OFFER = 'order-offers-v1';

/** Action ids. Must match the notifee actions and the iOS category actions. */
export const PUSH_ACTIONS = {
  DISMISS: 'dismiss',
  DETAILS: 'details',
} as const;

export type PushActionId = (typeof PUSH_ACTIONS)[keyof typeof PUSH_ACTIONS];

/** The `data` map as it crosses the wire. All values are strings. */
export interface PushPayload {
  v: string;
  type: string;
  order_id: string;
  route: string;
  title: string;
  body: string;
  /** ISO-8601. After this the notification is stale and self-cancels. */
  expires_at: string;
  category: string;
  /** `'1'` when the payload must be handled without being displayed. */
  silent?: string;
}

/** A payload that passed validation, with values usable by the app. */
export interface ParsedPush {
  type: PushType;
  orderId: string;
  route: string;
  title: string;
  body: string;
  expiresAt: Date | null;
  category: string;
  silent: boolean;
}

function isKnownType(value: string): value is PushType {
  return (Object.values(PUSH_TYPES) as string[]).includes(value);
}

/**
 * Validate and normalise. `null` means "ignore this message", never "error" —
 * a future server will send types this build has never heard of.
 */
export function parsePushPayload(data: unknown): ParsedPush | null {
  if (!data || typeof data !== 'object') return null;
  const raw = data as Record<string, unknown>;

  const str = (key: string): string =>
    typeof raw[key] === 'string' ? (raw[key] as string) : '';

  if (str('v') !== PUSH_CONTRACT_VERSION) return null;

  const type = str('type');
  if (!isKnownType(type)) return null;

  const orderId = str('order_id');
  if (!orderId) return null;

  const expiresRaw = str('expires_at');
  const expiresAt = expiresRaw ? new Date(expiresRaw) : null;

  return {
    type,
    orderId,
    route: str('route'),
    title: str('title'),
    body: str('body'),
    expiresAt: expiresAt && !Number.isNaN(expiresAt.getTime()) ? expiresAt : null,
    category: str('category') || PUSH_CATEGORY_OFFER,
    silent: str('silent') === '1',
  };
}

export function isExpired(push: ParsedPush, now: Date = new Date()): boolean {
  return push.expiresAt !== null && push.expiresAt.getTime() <= now.getTime();
}